### PR TITLE
fix: Check the device service existence before calling the device validation

### LINKED ISSUE #4478: Status code should return 207 instead of 503 when post/patch devices with not running device-service
# 🐞 Bug Report

### Affected Services [**REQUIRED**]

 The issue is located in: Metadata

### Is this a regression?

 Yes

### Description and Minimal Reproduction [**REQUIRED**]
1. Deploy EdgeX by [compose file](https://github.com/edgexfoundry/edgex-compose/blob/d3abd814d7ee08bb719e32901cdf7f306f4b71a8/docker-compose-no-secty.yml)
2. Post device and the serviceName is not created, example as below
```
[
    {
      "device":{
         "origin":1471806386919,
         "name":"Test-Device",
         "description":"Test-Device",
         "adminState":"UNLOCKED",
         "operatingState":"UP",
         "protocols":{
            "other":{
               "Address":"simple01",
               "Port":"300"
            }
         },
         "labels":[
            "device-sample"
         ],
         "location":"{lat:45.45,long:47.80}",
         "serviceName":"device-test",
         "profileName":"Random-UnsignedInteger-Device"
      },
      "apiVersion":"v2"
   }
]
```

## 🔥 Exception or Error
<pre><code>

Status code returns 503 with message "HTTP request timeout"
</code></pre>
Expect to return 207 on outside level with 503 status code inside 

## 🌍  Your Environment
**Deployment Environment:** Ubuntu / x86

**EdgeX Version [**REQUIRED**]:** Main

**Anything else relevant?**

### PR DESCRIPTION
**Pull request by jinlinGuan**
close #4478

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

1. Build core metadata image from this branch and run with other services
2. post/patch device with non-existence device service name
  ```
  curl -ki -X POST http://localhost:59881/api/v2/device -d '[
  {
    "apiVersion": "v2",
    "device": {
      "name": "test-1",
      "adminState": "UNLOCKED",
      "operatingState": "UP",
      "serviceName": "device-test",
      "profileName": "Random-Boolean-Device",
      "protocols": {
      "other": {}
      }
    }
  }]'
  ```
3. Verified the status code is 400 inside 207 Multi-Status
  ```
  HTTP/1.1 207 Multi-Status
  Content-Type: application/json
  X-Correlation-Id: 1a91f124-0f2f-4c5c-98f2-b1ab8d374683
  Date: Mon, 10 Apr 2023 07:15:13 GMT
  Content-Length: 96
  
  [{"apiVersion":"v2","message":"device service 'device-test' does not exists","statusCode":400}]
  ```

### CODE REVIEW

**Review by cloudxxx8 — APPROVED** [review]
LGTM

File: internal/core/metadata/application/device.go [hunk 1]
```diff
@@ -39,6 +39,14 @@ func AddDevice(d models.Device, ctx context.Context, dic *di.Container) (id stri
 	dbClient := container.DBClientFrom(dic.Get)
 	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
 
+	// Check the existence of device service before device validation
+	exists, edgeXerr := dbClient.DeviceServiceNameExists(d.ServiceName)
+	if edgeXerr != nil {
+		return id, errors.NewCommonEdgeXWrapper(edgeXerr)
```
Comment (lenny-intel) [inline internal/core/metadata/application/device.go:45]: Add context to the error as is done below for Update.

File: internal/core/metadata/controller/http/device_test.go [hunk 2]
```diff
@@ -539,16 +542,17 @@ func TestPatchDevice(t *testing.T) {
...
+	notFoundService.Device.ServiceName = &notFoundServiceName
+	dbClientMock.On("DeviceServiceNameExists", *notFoundService.Device.ServiceName).Return(false, nil)
 
 	notFountProfileName := "notFoundProfile"
 	notFoundProfile := testReq
 	notFoundProfile.Device.ProfileName = &notFountProfileName
 	notFoundProfileDeviceModel := dsModels
 	notFoundProfileDeviceModel.ProfileName = notFountProfileName
-	dbClientMock.On("DeviceById", *notFoundProfile.Device.Id).Return(notFoundProfileDeviceModel, nil)
+	//dbClientMock.On("DeviceById", *notFoundProfile.Device.Id).Return(notFoundProfileDeviceModel, nil)
```
Comment (lenny-intel) [inline internal/core/metadata/controller/http/device_test.go:555]: Remove commented out code

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/metadata/controller/http/device_test.go (hunk 2 above)
Comment (jinlinGuan) [reply internal/core/metadata/controller/http/device_test.go]: removed

File: internal/core/metadata/application/device.go (hunk 1 above)
Comment (jinlinGuan) [reply internal/core/metadata/application/device.go]: added

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**